GeosCore/set_boundary_conditions_mod.F90

```diff
@@ -0,0 +1,168 @@
+!EOP
+!------------------------------------------------------------------------------
+!BOC
+CONTAINS
+!EOC
+!------------------------------------------------------------------------------
+!                  GEOS-Chem Global Chemical Transport Model                  !
+!------------------------------------------------------------------------------
+!BOP
+!
+! !IROUTINE: set_boundary_conditions
+!
+! !DESCRIPTION: Subroutine SET\_BOUNDARY\_CONDITIONS sets the boundary conditions
+!  using the boundary conditions read from HEMCO for nested-grid simulations.
+!\\
+!\\
+! !INTERFACE:
+!
+ SUBROUTINE Set_Boundary_Conditions( Input_Opt, State_Chm, State_Grid, RC )
+!
+! !USES:
+!
+   USE ErrCode_Mod,      ONLY : GC_SUCCESS, GC_Error
+   USE Input_Opt_Mod,    ONLY : OptInput
+   USE Species_Mod,      ONLY : Species, SpcConc
+   USE State_Chm_Mod,    ONLY : ChmState
+   USE State_Grid_Mod,   ONLY : GrdState
+   USE Time_Mod,         ONLY : TIMESTAMP_STRING
+!
+! !INPUT PARAMETERS:
+!
+   TYPE(OptInput),   INTENT(IN   )          :: Input_Opt  ! Input options
+   TYPE(GrdState),   INTENT(IN   )          :: State_Grid ! Grid State
+!
+! !INPUT/OUTPUT PARAMETERS:
+!
+   TYPE(ChmState),   INTENT(INOUT)          :: State_Chm  ! Chemistry State
+   INTEGER,          INTENT(INOUT)          :: RC         ! Failure or success
+!
+! !REMARKS:
+!  Split off from HEMCO code (Get\_Boundary\_Conditions) in order to be called
+!  more frequently throughout timesteps.
+!
+! !REVISION HISTORY:
+!  28 Jul 2023 - H.P. Lin    - Initial version
+!  See https://github.com/geoschem/geos-chem for complete history
+!EOP
+!------------------------------------------------------------------------------
+!BOC
+!
+! !LOCAL VARIABLES:
+!
+   INTEGER              :: I, J, L, N, NA     ! lon, lat, lev, spc indexes
+   CHARACTER(LEN=255)   :: LOC                ! routine location
+   CHARACTER(LEN=16)    :: STAMP
+
+   !=================================================================
+   ! SET_BOUNDARY_CONDITIONS begins here!
+   !=================================================================
+
+   ! Name of this routine
+   LOC = ' -> at Set_Boundary_Conditions (in GeosCore/set_boundary_conditions_mod.F90)'
+
+   ! Assume success
+   RC        = GC_SUCCESS
+
+   ! We only need to get boundary conditions if this is a nested-grid
+   ! simulation.  Otherwise the BoundaryCond field won't be allocated.
+   IF ( .not. State_Grid%NestedGrid ) RETURN
+
+   ! Ensure species array is in kg/kg as State_Chm%BoundaryCond is in kg/kg dry
+   IF ( TRIM(State_Chm%Spc_Units) /= 'kg/kg dry' ) THEN
+      IF ( Input_Opt%amIRoot ) THEN
+          WRITE(6,*) 'Unit check failure: Current units are ', State_Chm%Spc_Units, ', expected kg/kg dry'
+      ENDIF
+      CALL GC_Error( 'Unit check failure: Cannot apply nested-grid boundary conditions if units are not kg/kg dry. Your run may have failed previous to this error.', RC, LOC )
+      RETURN
+   ENDIF
+
+   ! Loop over advected species
```

Comment:
We can probably collapse the N and L loops for better efficiency.  I can add this.